From 7f8b91ac67210f9bf0640acd3c8f6d9e05fb7696 Mon Sep 17 00:00:00 2001
From: QBos07 <qubos@outlook.de>
Date: Thu, 1 May 2025 15:22:29 +0000
Subject: [PATCH] Allow for cross-section relaxation for SuperH

Signed-off-by: QBos07 <qubos@outlook.de>
---
 bfd/elf32-sh.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/bfd/elf32-sh.c b/bfd/elf32-sh.c
index 39a837ce..612b1581 100644
--- a/bfd/elf32-sh.c
+++ b/bfd/elf32-sh.c
@@ -587,10 +587,12 @@ sh_elf_relax_section (bfd *abfd, asection *sec,
 	{
 	  /* A local symbol.  */
 	  Elf_Internal_Sym *isym;
+	  /* The corresponding section*/
+	  asection *isec;
 
 	  isym = isymbuf + ELF32_R_SYM (irelfn->r_info);
-	  if (isym->st_shndx
-	      != (unsigned int) _bfd_elf_section_from_bfd_section (abfd, sec))
+	  isec = bfd_section_from_elf_index(abfd, isym->st_shndx);
+	  if (isec == NULL)
 	    {
 	      _bfd_error_handler
 		/* xgettext:c-format */
@@ -600,8 +602,8 @@ sh_elf_relax_section (bfd *abfd, asection *sec,
 	    }
 
 	  symval = (isym->st_value
-		    + sec->output_section->vma
-		    + sec->output_offset);
+		    + isec->output_section->vma
+		    + isec->output_offset);
 	}
       else
 	{
-- 
2.49.0

